File: aoc-py/solutions/day23.py
"""
Day 23: A Long Walk

https://adventofcode.com/2023/day/23
"""
__all__ = ('Day23',)

from typing import ClassVar
from collections import defaultdict

from ..solution import Solution

class Day23(Solution):
    NAME: ClassVar[str] = 'A Long Walk'

    def _get_neighbors(self, grid: list[str], row: int, col: int) -> dict[str, tuple[int, int]]:
        """Returns the 4-neighborhood tiles of (i, j)

        while ensuring it is not a maze wall ('#')
        and that it does not go out of the grid bounds
        """
        n_rows = len(grid)
        n_cols = len(grid[0])

        return {
            direction: (row, col)
            for direction, (row, col) in (
                ('>', (row, col + 1)),
                ('<', (row, col - 1)),
                ('v', (row + 1, col)),
                ('^', (row - 1, col)),
            )
            if row in range(n_rows)
            and col in range(n_cols)
            and grid[row][col] != '#'
        }

    def _hike(self, inp: str, *, slopes: bool) -> int:
        grid = inp.splitlines()

        start = (0, grid[0].index('.'))

        last_row = len(grid) - 1
        end = (last_row, grid[last_row].index('.'))

        nodes = [start, end]

        for i, row in enumerate(grid):
            for j, tile in enumerate(row):
                if tile != '#' and len(self._get_neighbors(grid, i, j)) >= 3:
                    # crossroad points, where we can make a choice of where to go
                    # they will be the nodes for the constructed graph
                    nodes.append((i, j))

        # graph of nodes (crossroads) + edges between them
        #
        # maps: node -> (map: connected nodes -> distance away from starting node)
        # only stores directly adjacent nodes
        graph = defaultdict(dict)

        for starting_node in nodes:
            to_check = [(starting_node, int())]
            seen = {starting_node}

            while to_check:
                # distance = edge length
                node, distance = to_check.pop()

                if distance > 0 and node in nodes:
                    # we've reached a crossroad/node, we can add it to the graph
                    graph[starting_node][node] = distance
                else:
                    # hit a regular tile
                    # keep on traversing, "floodfilling" through the neighbors of each tile
                    row, col = node
                    connected_nodes = self._get_neighbors(grid, row, col)

                    for node in (
                        [slope]
                        # we've reached a slope tile: '<', '>', '^', 'v'
                        # obtain the single direction (neighbor in that direction)
                        # that we are allowed to go to according to the slope
                        if slopes and (slope := connected_nodes.get(grid[row][col]))
                        # '.' tile, regular tile, check all 4 neighbors
                        else connected_nodes.values()
                    ):
                        if node not in seen:
                            # add tile as part of edge, increment distance
                            to_check.append((node, distance + 1))
                            seen.add(node)

        seen = set()

        def _dfs(node: tuple[int, int]) -> int:
            """Brutes force the length of the longest path

            depth-first-search traversal through all path possibilities using generated graph
            """
            if node == end:
                return 0
            # `seen` set to avoid cycles.
            seen.add(node)

            max_length = max([
                # recursively pathfinds, and adds up all the distances of the edges
                _dfs(next_node) + graph[node][next_node]
                for next_node in graph[node]
                if next_node not in seen
            ] + [0]) # `0` if no path found, avoids max() arg is an empty sequence error

            seen.remove(node)
            return max_length

        return _dfs(start)

    def part_one(self, inp: str) -> int:
        return self._hike(inp, slopes=True)

    def part_two(self, inp: str) -> int:
        return self._hike(inp, slopes=False)

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 2182
        assert p2 == 6670

File: src/bin/day23.rs
//! Day 23: A Long Walk
//!
//! <https://adventofcode.com/2023/day/23>
use std::{
    collections::{HashMap, HashSet, VecDeque},
    fmt::Display,
};
use aoc_2023::Solution;

pub struct Day23;

type Coord = (usize, usize);

impl Day23 {
    fn get_neighbors(
        grid: &[Vec<u8>],
        row: usize,
        col: usize,
    ) -> HashMap<u8, Coord> {
        [
            (b'>', (row, col + 1)),
            (b'<', (row, col.wrapping_sub(1))),
            (b'v', (row + 1, col)),
            (b'^', (row.wrapping_sub(1), col)),
        ]
        .into_iter()
        .filter(|&(_, (i, j))| grid
            .get(i)
            .and_then(|row| row.get(j))
            .map(|&tile| tile != b'#')
            .unwrap_or_default()
        )
        .collect::<HashMap<u8, Coord>>()
    }

    fn dfs(
        node: Coord,
        end: Coord,
        graph: &HashMap<Coord, HashMap<Coord, usize>>,
        seen: &mut HashSet<Coord>,
    ) -> usize {
        if node == end { 0 }
        else {
            seen.insert(node);

            #[allow(clippy::filter_map_bool_then)]
            let max_length = graph
                .get(&node)
                .map_or(0, |connected_nodes|
                    connected_nodes
                        .keys()
                        .filter_map(|next_node|
                            (!seen.contains(next_node))
                                .then(|| Self::dfs(*next_node, end, graph, seen) + graph[&node][&next_node])
                        )
                        .max()
                        .unwrap_or(0)
                );

            seen.remove(&node);
            max_length
        }
    }

    fn hike<T: Display>(inp: T, slopes: bool) -> usize {
        let grid = inp
            .to_string()
            .lines()
            .map(|line| line
                .as_bytes()
                .to_vec()
            )
            .collect::<Vec<Vec<u8>>>();

        let start = (0, grid[0]
            .iter()
            .position(|&c| c == b'.')
            .unwrap()
        );
        let end = {
            let last_row = grid.len() - 1;
            (last_row, grid[last_row]
                .iter()
                .position(|&c| c == b'.')
                .unwrap()
            )
        };

        let mut nodes = vec![start, end];
        let grid_ref = &grid;

        nodes.extend(
            grid.iter()
                .enumerate()
                .flat_map(|(i, row)| row
                    .iter()
                    .enumerate()
                    .filter_map(move |(j, &tile)|
                        (tile != b'#' && Self::get_neighbors(grid_ref, i, j)
                            .len() >= 3
                        )
                        .then_some((i, j))
                    )
                )
        );

        let mut graph = HashMap::new();

        for starting_node in &nodes {
            let mut to_check = VecDeque::from([
                (*starting_node, 0)
            ]);
            let mut seen = HashSet::new();

            while let Some((node @ (row, col), distance)) = to_check.pop_front() {
                if distance > 0 && nodes.contains(&node) {
                    graph
                        .entry(*starting_node)
                        .or_insert_with(HashMap::new)
                        .insert(node, distance);
                } else {
                    let connected_nodes = Self::get_neighbors(grid_ref, row, col);
                    let next_nodes =
                        if let Some(&slope) = grid
                            .get(row)
                            .and_then(|row| row.get(col))
                            .and_then(|node| connected_nodes.get(node))
                            .filter(|_| slopes)
                        { vec![slope] }
                        else { connected_nodes
                            .into_values()
                            .collect::<Vec<Coord>>()
                        };

                    for node in next_nodes {
                        if !seen.contains(&node) {
                            to_check.push_back((node, distance + 1));
                            seen.insert(node);
                        }
                    }
                }
            }
        }
        Self::dfs(start, end, &graph, &mut HashSet::new())
    }

    pub fn part_one<T: Display>(&self, inp: T) -> usize {
        Self::hike(inp, true)
    }

    pub fn part_two<T: Display>(&self, inp: T) -> usize {
        Self::hike(inp, false)
    }
}

impl Solution for Day23 {
    const NAME: &'static str = "A Long Walk";

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 2182);
        assert_eq!(p2, 6670);
    }
}

fn main() {
    aoc_2023::run_day(23, &Day23);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}